yew/src/dev/messages.rs

```diff
@@ -0,0 +1,125 @@
+//! Messages describing the state of an application.
+
+use cfg_if::cfg_if;
+use serde::Serialize;
+
+cfg_if! {
+    if #[cfg(feature = "std_web")] {
+        use stdweb::web::{Node, INode};
+    } else if #[cfg(feature = "web_sys")] {
+        use web_sys::Node;
+    }
+}
+
+/// Sent when something happens to a component.
+#[derive(Serialize, Debug)]
+pub enum ComponentEvent {
+    /// Sent when a component mounts to the DOM
+    Mounted,
+    /// Sent when a component unmounts from the DOM
+    Unmounted,
+    /// Sent when a component updates itself
+    Updated,
+    /// Sent when a component is created
+    Created,
+    /// Sent when a component is destroyed
+    Destroyed,
+}
+
+/// Generates a selector for a given node which can then be sent over a WebSocket
+/// connection to the browser extension.
+pub fn selector(dom_node: &Node) -> String {
+    let mut current_node = Some(dom_node.clone());
+    let mut selector_string = String::new();
+    loop {
+        match current_node {
+            Some(node) => {
+                cfg_if! {
+                    if #[cfg(feature="std_web")] {
+                        selector_string = node.node_name() + "/" + &selector_string;
+                        current_node = node.parent_node();
+                    } else if #[cfg(feature="web_sys")] {
+                        selector_string = node.node_name() + "/" + &selector_string;
+                        current_node = node.parent_node();
+                    }
+                }
+            }
+            None => {
+                return selector_string;
+            }
+        };
+    }
+}
+
+/// Stores data about a component (currently only the name of the struct).
+#[derive(Serialize, Debug)]
+pub struct DebugComponent {
+    /// The name of the component
+    name: String,
+    selector: Option<String>,
+}
+
+impl DebugComponent {
+    /// Creates a new instance of `DebugComponent`
+    pub fn new(name: String, selector: Option<String>) -> Self {
+        Self { name, selector }
+    }
+}
+
+/// A message sent to describe a change in a component's state.
+#[derive(Serialize, Debug)]
+pub struct ComponentMessage {
+    /// Time in seconds since the page was loaded.
+    time: f64,
```

Comment:
What is the purpose of sending this?

Reply:
I thought it could be used to build a timeline of events on the page.

Reply:
The extension will receive messages in order so I think it can handle timestamping on its end

Reply:
That makes sense, I'll remove the field.